File: problem_set_1/palindrome_pairs.go
package main

import "fmt"

func reverseString(s string) string {
	var reverse = ""
	var start = len(s) - 1

	for i := start; i >= 0; i-- {
		reverse = reverse + string(s[i])
	}

	return reverse; 
}

func isPalindrome(s string) bool {
	var reverse = reverseString(s) 

	if s == reverse {
		return true
	}

	return false
}

func palindromePairs(words []string) [][]int {
	var pairs [][]int 

	for i := 0; i < len(words); i++ {
		for j:= 0; j < len(words); j++ {
			if i != j {
				var concatinated = words[i] + words[j]
				if isPalindrome(concatinated) {
					var pair = []int{i, j}
					pairs = append(pairs, pair) 
				}
			}
		}
	}

	return pairs; 
}

func main() {
	inputWords := []string{"bat", "tab", "cat"}
	result := palindromePairs(inputWords)
	fmt.Println(result) // Output: [[0 1] [1 0]]
}

File: problem_set_1/PalindromePairs.java
import java.util.*; 

public class PalindromePairs {
    static boolean isPalindrome(String s) {
        StringBuilder reverse = new StringBuilder(s); 
        reverse.reverse(); 

        if(s.equals(reverse.toString())) {
            return true; 
        }

        return false; 
    }

    static ArrayList<int[]> palindromePairs(String[] words) {
        ArrayList<int[]> pairs = new ArrayList<>(); 

        for(int i = 0; i < words.length; i++) {
            for(int j = 0; j < words.length; j++) {
                if(i != j) {
                    String concatinated = words[i] + words[j]; 

                    if(isPalindrome(concatinated)) {
                        int[] pair = {i, j}; 
                        pairs.add(pair); 
                    }
                }
            }
        }

        return pairs; 
    }

    public static void main(String[] args) {
        String[] words = {"tab", "bat", "cat"}; 

        ArrayList<int[]> list = palindromePairs(words); 

        for(int[] pair: list) {
            System.out.print(Arrays.toString(pair) + ", ");
        }
    }
}